### DISCUSSION TITLE
Error when using `ensureQueryData` with `infiniteQueryOptions`

### DISCUSSION THREAD

**Discussion by chinanderm** [TanStack/query#6712] [Q&A] [answered]
I'm having issues in v5 when trying to use `ensureQueryData()` with `infiniteQueryOptions()` and `useSuspenseInfiniteQuery()`.

See this StackBlitz for a reproduction: https://stackblitz.com/edit/vitejs-vite-tnfvrc?file=src%2Froutes%2Findex.tsx

In a TanStack Router `loader` function I am using `queryClient.ensureQueryData`, passing in the result of a `infiniteQueryOptions()` call. The initial data fetch is successful but in the route component when calling `useSuspenseInfiniteQuery()` with the `infiniteQueryOptions()` result, an error is getting thrown: `Cannot read properties of undefined (reading 'length')`, which appears to be coming from `getNextPageParam` internal function. In this location, `pages` is `undefined`.
![Screenshot 2024-01-15 at 11 24 37 AM](https://github.com/TanStack/query/assets/3341414/595a79a4-acaf-4770-8a27-3f2162f942ab)

Am I composing things incorrectly? Are `infiniteQueryOptions()` and `useSuspenseInfiniteQuery()` incompatible with `ensureQueryData()`?

Also, despite setting `initialPageParam`, the `pageParam` value passed into the `queryFn` is `undefined` on the first invocation. Why is that?

**Comment by TkDodo** [comment]
`ensureQueryData` calls `fetchQuery` under the hood, but for infinite queries, it would need to call `fetchInfiniteQuery`.

Since it's just a thin wrapper over `getQueryData ?? fetchQuery`, you can implement it in user-land pretty quickly:

```
const data =
  queryClient.getQueryData(myInfiniteQueryOptions().queryKey) ??
  (await queryClient.fetchInfiniteQuery(myInfiniteQueryOptions()));
```

> Also, despite setting initialPageParam, the pageParam value passed into the queryFn is undefined on the first invocation. Why is that?

can't confirm that, probably a side-effect of the above issue ?

forked, working stackblitz: https://stackblitz.com/edit/vitejs-vite-suancq

**Reply by chinanderm** [reply]
Thank you @TkDodo -- that worked and resolved the `pageParam` being `undefined` issue.

**Reply by mrlubos** [reply]
@TkDodo are there any plans to add this functionality to the router? e.g. `ensureInfiniteQueryData()`?

**Reply by mrlubos** [reply]
Side note, it seems your example has a type problem in `useSuspenseInfiniteQuery()`

**Reply by mrlubos** [reply]
@TkDodo hmm, how would you fetch 2+ pages in your example? Since cursor isn't a part of the query key... https://stackblitz.com/edit/vitejs-vite-suancq?file=src%2Froutes%2Findex.tsx%3AL34-L36

**Reply by TkDodo** [reply]
> ensureInfiniteQueryData

Yeah I guess we could / should

> how would you fetch 2+ pages 

you can pass `pages: 2` to `fetchInfiniteQuery` to fetch two pages instead of one

**Reply by mrlubos** [reply]
Hmm @TkDodo, is there any other way for the second question? My use case is something like loading files in a GitHub pull request when you visit a direct link. I don't know how many pages I need to load before I have loaded the target file and can render the list, and I don't want to load everything for obvious reasons

**Reply by TkDodo** [reply]
not entirely sure what you mean. what decides how many pages you want to load?

**Reply by mrlubos** [reply]
@TkDodo I've got a URL containing an optional ID. If it's not present, I'll load the first page. If it's present, I keep loading pages one by one until I run out of pages or I find the ID in page results, whatever happens first. This logic currently lives inside the component and I'm moving it to loader, that's why I run into this

**Reply by mrlubos** [reply]
@TkDodo to add, I'm able to fetch pages as described in my logic, but only if I make page param part of the query key. That's most likely not what I want as each page is its own query. If I'm passing updated page param in initialPageParam, I only get the first page/fetch and every subsequent attempt is ignored as there's cached data (that's what I think is happening)

**Reply by mrlubos** [reply]
I tried messing about with initial data and placeholder data thinking I could keep seeding the cache, but it all gets discarded, so I got stuck and came here haha

**Reply by TkDodo** [reply]
you can set `pages` to a very high number or `Infinity` to keep fetching until `getNextPageParam` returns `undefined`

**Reply by mrlubos** [reply]
@TkDodo Would I be able to implement the "check for ID in a page" logic inside getNextPageParam and return undefined if it's found? I mean, I'm sure I could, but is it advisable?

**Reply by TkDodo** [reply]
sure

**Reply by mrlubos** [reply]
@TkDodo setting `pages` to a high number will do in my case, thanks! However, it's not perfect. Here's an example https://stackblitz.com/edit/vitejs-vite-w9jro4?file=src%2Froutes%2Findex.tsx

Note that I am asking for 10 pages, but if page is larger than 2, I return undefined. This would mean that there are no more pages, and we should abandon attempting to fetch more. However, you can see in the console that all request attempts were made. This is wasteful and requires a balance between setting a good limit and app performance. If I set it to `Infinity` as suggested, that does NOT work because `infiniteQueryOptions()` will simply keep running forever. Do you think this should be reported as a bug? I think I understand what it's trying to do, but not sure that's a correct behaviour

**Reply by TkDodo** [reply]
thanks for the reproduction; we do have a test that only the amount of pages winds up in the cache, but wer're not eagerly exiting the loop here:

https://github.com/TanStack/query/blob/7edf587418b0dcc240c5a5be22e626b01b4c7cd1/packages/query-core/src/infiniteQueryBehavior.ts#L103-L107

I think this is a bug

**Reply by TkDodo** [reply]
here's the fix:
- https://github.com/TanStack/query/pull/7799

**Comment by TkDodo** [comment, marked as answer]
We’re now shipping a dedicated `ensureInfiniteQueryData` function for this:

https://tanstack.com/query/latest/docs/reference/QueryClient#queryclientensureinfinitequerydata

**Reply by mrlubos** [reply]
Very cool!
